Read csv headers

cdknox opened the pull request:
- [x] closes #36688 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
storage_options as headers and tests added

Commit:
additional tests - gzip, test additional headers receipt

Commit:
bailed on using threading for testing

Commit:
clean up comments add json http tests

Commit:
Merge branch 'master' into read_csv_headers to update

Commit:
added documentation on storage_options for headers

Commit:
DOC:Added doc for custom HTTP headers in read_csv and read_json

jreback commented:
cc @twoertwein

jreback commented on doc/source/user_guide/io.rst:
```diff
@@ -1625,6 +1625,18 @@ functions - the following example shows reading a CSV file:
 
    df = pd.read_csv("https://download.bls.gov/pub/time.series/cu/cu.item", sep="\t")
 
+A custom header can be sent alongside HTTP(s) requests by passing a dictionary
```
add versionadded 1.2 tag

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -221,6 +221,23 @@ Additionally ``mean`` supports execution via `Numba <https://numba.pydata.org/>`
+Custom HTTP(s) headers when reading csv or json files
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:meth:`read_csv` and :meth:`read_json` use the dictionary passed to ``storage_options`` to create custom HTTP(s) headers.
```
add the issue number.

Commit:
DOC:Corrected versionadded tag and added issue number for reference

twoertwein commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +414,151 @@ def test_is_fsspec_url():
     assert not icom.is_fsspec_url("relative/local/path")
+
+
+def test_plain_text_read_csv_http_custom_headers():
```
it is probably good to add a testcase for `read/to_parquet` as well as they have a slightly different IO handling (using `get_handle` only if needed).

twoertwein commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +414,151 @@ def test_is_fsspec_url():
+                "http://localhost:80/test.json", storage_options=headers
+            )
+            assert dummy_request.called_with(headers=headers)
+            assert (received_df == true_df).all(axis=None)
```
tm.assert_frame_equal

twoertwein commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +414,151 @@ def test_is_fsspec_url():
+                "http://localhost:80/test.json", storage_options=headers
+            )
+            assert dummy_request.called_with(headers=headers)
+            assert (received_df == true_df).all(axis=None)
```
tm.assert_frame_equal

twoertwein reviewed (commented):
Great work! The only part missing is the documentation update in pandas/core/shared_docs.py and making sure that all functions that take `storage_options` work as expected with it. imho I would only add a new test for parquet, the other functions forward everything directly to `get_handle`.

Commit:
DOC:updated storage_options documentation

Commit:
TST:updated with tm.assert_frame_equal

Commit:
TST:fixed incorrect usage of tm.assert_frame_equal

Commit:
CLN:reordered imports to fix pre-commit error

Commit:
DOC:changed whatsnew and added to shared_docs.py GH36688

Commit:
ENH: read nonfsspec URL with headers built from storage_options GH36688

Commit:
TST:Added additional tests parquet and other read methods GH36688

twoertwein commented on pandas/core/shared_docs.py:
```diff
@@ -383,8 +383,11 @@
-    will be raised if providing this argument with a non-fsspec URL.
-    See the fsspec and backend storage implementation docs for the set of
-    allowed keys and values."""
+    host, port, username, password, etc. If using a URL that will
```
Maybe something a long the lines of
"[... host, port, username, password, etc.] Key-value pairs for HTTP(S) URLs are forwarded to``urllib`` as header options. For other URLs , e.g., starting with "s3://" and "gcs://", key-value pairs are forwarded to ``fsspec``. Please see ``fsspec`` and ``urllib`` for more details.

cdknox replied:
Sounds reasonable enough to me.  Unrelated to the documentation change (but something I'll address in the next commit if you could provide guidance) I noticed that some of the parquet tests I wrote are failing on the checks that get run due to there not being a usable engine.  Am I missing something there or should I try to write tests that don't rely on having a parquet parsing engine?  Also, it looks like we don't like unittest and prefer pytest with monkeypatching, but I can't seem to find a MagicMock equivalent in pytest.  Would we rather have the import or should I write my own class for that?  Thanks for the guidance!

twoertwein replied:
About the docs, maybe even replace `fsspec` with `fsspec.open(**storage_options)` and `urllib` with `urllib.request.Request(headers=storage_options)` but that might be too verbose, idk.

Parquet needs optional dependencies, either `pyarrow` of `fastparquet`. It would be good to parametrize your test to cover both engines (if both engines support the new feature), then you can easily add a `pytest.importorskip`.

Sorry, I'm not familiar with monkeypatching/MagicMock.

jreback commented:
we have tons of fixtures eg s3_resources

pls try to not reinvent the wheel 

if u must mock then use monkeypatch

prefer not to mock at all

ivanovmg commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +413,297 @@ def test_is_fsspec_url():
+        "Auth": "other_custom",
+    }
+
+    class DummyResponse:
+        headers = {
+            "Content-Type": "text/csv",
+        }
+
+        @staticmethod
+        def read():
+            return df_csv_bytes
+
+        @staticmethod
+        def close():
+            pass
```
This class looks similar in all cases but one.
Can it be parametrized?

```
class DummyResponse:
    def __init__(self, data):
        self.data = data

    def read(self):
        return self.data
```

Then only function ``dummy_response_getter`` should be updated accordingly.

ivanovmg replied:
My bad, I noticed that ``headers`` class attribute is different from case to case.
But that _can_ be parametrized too (not that I insist on it).

cdknox replied:
I'll have a go at parameterizing it, getting back to this today.  That said, after Jeff's comment on preferring not to mock (in combination with not being sure how to use monkeypatch in a way that legitimately tests this functionality) I think I will rewrite the tests to use threading and a simple http server like I had in a past commit.  I bailed on that in favor of patching because I seemed to need a port per test and wasn't sure if I could rely on ports not being taken.  I'm fine with that route though if that's what's preferred, any guidance offered (ports, useful fixtures that already exist, etc.) is well taken.

Commit:
TST:removed mocking in favor of threaded http server

Commit:
DOC:refined storage_options doscstring

cdknox committed:
Merge branch 'master' into read_csv_headers

pep8speaks commented:
Hello @cdknox! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/io/parquet.py`](https://github.com/pandas-dev/pandas/blob/018a3990b501beb66689429a6d74f808c951b672/pandas/io/parquet.py):

> [Line 420:1](https://github.com/pandas-dev/pandas/blob/018a3990b501beb66689429a6d74f808c951b672/pandas/io/parquet.py#L420): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 482:64](https://github.com/pandas-dev/pandas/blob/018a3990b501beb66689429a6d74f808c951b672/pandas/io/parquet.py#L482): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 485:1](https://github.com/pandas-dev/pandas/blob/018a3990b501beb66689429a6d74f808c951b672/pandas/io/parquet.py#L485): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

Commit:
CLN:used the github editor and had pep8 issues

Commit:
CLN: leftover comment removed

Commit:
TST:attempted to address test warning of unclosed socket GH36688

Commit:
TST:added pytest.importorskip to handle the two main parquet engines GH36688

Commit:
CLN: imports moved to correct order GH36688

Commit:
TST:fix fastparquet tests GH36688

Commit:
CLN:removed blank line at end of docstring GH36688

Commit:
CLN:removed excess newlines GH36688

Commit:
CLN:fixed flake8 issues GH36688

Commit:
TST:renamed a test that was getting clobbered and fixed the logic GH36688

Commit:
CLN:try to silence mypy error via renaming GH36688

Commit:
TST:pytest.importorfail replaced with pytest.skip GH36688

Commit:
TST:content of dataframe on error made more useful GH36688

Commit:
CLN:fixed flake8 error GH36688

Commit:
TST: windows fastparquet error needs raised for troubleshooting GH36688

Commit:
CLN:fix for flake8 GH36688

Commit:
TST:changed compression used in to_parquet from 'snappy' to None GH36688

The default compression on to_parquet is not installed on one of the
test environments. It should work just the same with no compression anyway.

Commit:
TST:allowed exceptions to be raised via removing a try except block GH36688

twoertwein commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +414,316 @@ def test_is_fsspec_url():
+)
+def test_server_and_default_headers(responder, read_method, port, parquet_engine):
+    if parquet_engine is not None:
+        if parquet_engine == "pyarrow":
```
Would `pytest.importorskip` work here instead of try-except?

cdknox replied:
I had that in a prior commit and it seemed like it still ran the test, so I replaced it with a try-except.  It must have been confounded with something else I hadn't quite worked out.  I'll give that a go again with ``pytest.importorskip``, sorry about that!

Commit:
TST:replaced try except with pytest.importorskip GH36688

cdknox commented:
I'm sure folks are busy; I just wanted to make sure no one is waiting on me for anything.  I think I've addressed the comments so far, but if I'm missing something or I need to do something else please let me know!

twoertwein commented on pandas/io/common.py:
```diff
@@ -276,12 +276,17 @@ def _get_filepath_or_buffer(
+        # TODO: fsspec can also handle HTTP via requests, but leaving this
+        # unchanged. using fsspec appears to break the ability to infer if the
+        # server responded with gzipped data
+        storage_options = storage_options or dict()
```
`storage_options = storage_options or {}`
I think there is an open issue about avoiding calls to `dict`

twoertwein commented on pandas/io/common.py:
```diff
@@ -276,12 +276,17 @@ def _get_filepath_or_buffer(
+        import urllib.request
+
+        # assuming storage_options is to be interpretted as headers
+        req_info = urllib.request.Request(filepath_or_buffer, headers=storage_options)
```
are there more (important) keyword arguments that `Request` takes? If yes, it might be more future-proof to do:
`urllib.request.Request(filepath_or_buffer, **storage_options)`

cdknox replied:
I like that thought, though from reviewing the options I don't think there's much to be gained by exposing the other keyword arguments.  It looks like the other kwargs are there mainly to post data to a url, but that doesn't look to be possible inside the ``_get_filepath_or_buffer`` function due to what is available when that function is called.  Here's the documentation link in case anyone has a different take: [https://docs.python.org/3/library/urllib.request.html#urllib.request.Request](url)

Another less restrictive option might be to change it so that ``storage_options = {"headers": {"User-Agent": "pandas"}}`` leading to ``urllib.request.Request(filepath_or_buffer, headers=storage_options["headers"])``.  This means the options could be used for other functions in ``_get_filepath_or_buffer``.

Both may be a little cumbersome on the user, since I see little benefit in this function in the http(s) case, but it may provide some flexibility that would be nice elsewhere.  Easy enough to do either option and change the docs, or leave it as is.  Got any particular preference among those options?  Thanks!

twoertwein replied:
`is_url` strings that are not `is_fsspec_url`, did not use any storage_options before this PR. All options could go into the header/to urllib.request.Request.
If we do something like `Request(filepath_or_buffer, headers=storage_options["headers"])` then we could have other options that are not used but we don't raise about them (or urllib doesn't complain about the provided arguments).

cdknox replied:
As far as I know, there are common uses for some header keys but there's also not really a limit on what header keys can be used (ie. I think all keys are accepted in headers).  To back this up the ``urllib.request.Request`` documentation says that it will call the ``urllib.request.Request.add_header`` method on each key value pair, and from looking at that method and the ``__init__`` method, I don't see any restriction on keys allowed in the header dictionary either.

I think once I am free later tonight I will go ahead and change it so that the usage and documentation lines up with ``Request(filepath_or_buffer, headers=storage_options["headers"])``

Commit:
CLN:removed dict() in favor of {} GH36688

jreback commented on doc/source/user_guide/io.rst:
```diff
@@ -1625,6 +1625,20 @@ functions - the following example shows reading a CSV file:
 
    df = pd.read_csv("https://download.bls.gov/pub/time.series/cu/cu.item", sep="\t")
 
+.. versionadded:: 1.2.0
```
would need to be 1.3

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -251,6 +251,26 @@ Additionally ``mean`` supports execution via `Numba <https://numba.pydata.org/>`
 the  ``engine`` and ``engine_kwargs`` arguments. Numba must be installed as an optional dependency
 to use this feature.
 
+.. _whatsnew_120.read_csv_json_http_headers:
```
would need to be 1.3

jreback commented on pandas/io/parquet.py:
```diff
@@ -404,10 +416,12 @@ def to_parquet(
     path,
     engine: str = "auto",
     columns=None,
+    storage_options: StorageOptions = None,
```
need Optional here

twoertwein replied:
`StorageOptions` is already defined with an `Optional`: `StorageOptions = Optional[Dict[str, Any]]`

jreback replied:
ahh ok

jreback commented on pandas/io/parquet.py:
```diff
@@ -432,13 +446,18 @@ def read_parquet(
+
+    {storage_options}
+
+        .. versionadded:: 1.2.0
```
1.3

jreback commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +414,298 @@ def test_is_fsspec_url():
+            pass
+
+        bio = BytesIO()
+        bio.close = dummy_close
```
do you need to do this in a mock? / monkeypatch?

cdknox replied:
I found that fsspec allowed for "writing to a file" in memory.  I changed to using that for creating the fast parquet bytes reponse which avoids any patching/mocking here.

jreback commented on pandas/tests/io/test_common.py:
```diff
@@ -411,3 +414,298 @@ def test_is_fsspec_url():
     assert not icom.is_fsspec_url("random:pandas/somethingelse.com")
     assert not icom.is_fsspec_url("/local/path")
     assert not icom.is_fsspec_url("relative/local/path")
+
```
can you make a new test file for all of these additions. test_user_agent.py or similar

Commit:
Merge branch 'master' into read_csv_headers

Commit:
DOC: changed potentially included version from 1.2.0 to 1.3.0 GH36688

Commit:
TST:user agent tests moved from test_common to their own file GH36688

Commit:
TST: used fsspec instead of patching bytesio GH36688

Commit:
TST: added importorskip for fsspec on FastParquet test GH36688

Commit:
TST:added missing importorskip to fsspec in another test GH36688

twoertwein commented on pandas/io/common.py:
```diff
@@ -276,12 +276,18 @@ def _get_filepath_or_buffer(
+
+        # waiting until now for importing to match intended lazy logic of
+        # urlopen function defined elsewhere in this module
+        import urllib.request
```
the other libraries that are imported in `_get_filehandle_or_buffer` are optional dependencies (fsspec and botocore). urllib is a standard library that should be available in all python versions supported by pandas 1.2. Would make sense to move it to the top.

cdknox commented on pandas/tests/io/test_user_agent.py:
```diff
@@ -0,0 +1,309 @@
+
+import pandas as pd
+import pandas._testing as tm
+
```
I think the confounding thing here is that not having fsspec installed breaks a test, and not the actual feature being tested (pandas could set custom http headers on a FastParquet file request and parse it without fsspec being installed).  The fsspec package is only needed in the case of testing FastParquet because it won't write to a BytesIO or filelike object without being patched.  I didn't importorskip in the class definition because it's outside of a test, and used inside a test.  I don't know if that helped explain it or made it less clear.  If you'd rather it be done another way, that's all good just let me know!

jreback commented:
thanks @cdknox very nice

cdknox commented:
Glad to help as a long time pandas user! Just double checking, we're good without the change then from ``storage_options={"User-Agent": "custom_user_agent"}`` to ``storage_options={"headers": {"User-Agent": "custom_user_agent"}}``?  I'm good with it either way

simonjayhawkins commented on pandas/io/parquet.py:
```diff
@@ -404,10 +416,12 @@ def to_parquet(
     path,
     engine: str = "auto",
     columns=None,
+    storage_options: StorageOptions = None,
```
This new parameter should maybe be added after use_nullable_dtypes
